### PR TITLE
[SPARK-32509][SQL] Ignore unused DPP True Filter in Canonicalization

### PR DESCRIPTION
**Pull request by prakharjain09**
### What changes were proposed in this pull request?
This PR fixes issues relate to Canonicalization of FileSourceScanExec when it contains unused DPP Filter.

### Why are the changes needed?

As part of PlanDynamicPruningFilter rule, the unused DPP Filter are simply replaced by `DynamicPruningExpression(TrueLiteral)` so that they can be avoided. But these unnecessary`DynamicPruningExpression(TrueLiteral)` partition filter inside the FileSourceScanExec affects the canonicalization of the node and so in many cases, this can prevent ReuseExchange from happening.

This PR fixes this issue by ignoring the unused DPP filter in the `def doCanonicalize` method. 

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Added UT.

### CODE REVIEW

**Comment by prakharjain09** [comment]
cc - @cloud-fan @maryannxue

**Comment by cloud-fan** [comment]
ok to test

File: sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala [hunk 1]
```diff
@@ -608,12 +608,16 @@ case class FileSourceScanExec(
...
+    predicates.filterNot(_ == Literal.TrueLiteral)
+  }
+
   override def doCanonicalize(): FileSourceScanExec = {
     FileSourceScanExec(
       relation,
       output.map(QueryPlan.normalizeExpressions(_, output)),
       requiredSchema,
-      QueryPlan.normalizePredicates(partitionFilters, output),
+      QueryPlan.normalizePredicates(filterUnnecessaryPredicates(partitionFilters), output),
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala:620]: Why does `TrueLiteral` affect canonicalize? `TrueLiteral == TrueLiteral` returns true, right?

Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala:620]: Ah I got it. Maybe one side has DPP and one side not.

 Can we filter out `DynamicPruningExpression(Literal.TrueLiteral)`? Then we don't need to change https://github.com/apache/spark/pull/29318/files#diff-6c00f36fadec282fce72ad39cfea51caL81 and make this PR less risky.

Comment (prakharjain09) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala]: done.

**Comment by prakharjain09** [comment]
@cloud-fan Thanks for the review. I have made the required changes.

**Comment by cloud-fan** [comment]
thanks, merging to master/3.0!

**Comment by prakharjain09** [comment]
@cloud-fan Thanks!

**Pull request closed without merging**